Source/NetworkReachabilityManager.swift

```diff
@@ -137,86 +151,100 @@ open class NetworkReachabilityManager {
-            reachability,
-            { (_, flags, info) in
-                let reachability = Unmanaged<NetworkReachabilityManager>.fromOpaque(info!).takeUnretainedValue()
-                reachability.notifyListener(flags)
-            },
-            &context
-        )
-
-        let queueEnabled = SCNetworkReachabilitySetDispatchQueue(reachability, listenerQueue)
-
-        listenerQueue.async {
-            guard let flags = self.flags else { return }
-            self.notifyListener(flags)
+    open func startListening(onQueue queue: DispatchQueue = .main,
+                             onUpdatePerforming listener: @escaping Listener) -> Bool {
+        stopListening()
+
+        listenerQueue = queue
+        self.listener = listener
+
+        var context = SCNetworkReachabilityContext(version: 0,
+                                                   info: Unmanaged.passRetained(self).toOpaque(),
+                                                   retain: nil,
+                                                   release: nil,
+                                                   copyDescription: nil)
+        let callback: SCNetworkReachabilityCallBack = { (target, flags, info) in
+            guard let info = info else { return }
+
+            let instance = Unmanaged<NetworkReachabilityManager>.fromOpaque(info).takeUnretainedValue()
+            instance.notifyListener(flags)
         }
 
-        return callbackEnabled && queueEnabled
+        let queueAdded = SCNetworkReachabilitySetDispatchQueue(reachability, reachabilityQueue)
+        let callbackAdded = SCNetworkReachabilitySetCallback(reachability, callback, &context)
+
+        // Manually call listener to give initial state, since the framework may not.
+        if let currentFlags = flags {
+            reachabilityQueue.async {
+                self.notifyListener(currentFlags)
+            }
+        }
+
+        return callbackAdded && queueAdded
     }
 
     /// Stops listening for changes in network reachability status.
     open func stopListening() {
         SCNetworkReachabilitySetCallback(reachability, nil, nil)
         SCNetworkReachabilitySetDispatchQueue(reachability, nil)
+        previousStatus = nil
+        listenerQueue = nil
+        listener = nil
     }
 
     // MARK: - Internal - Listener Notification
 
     func notifyListener(_ flags: SCNetworkReachabilityFlags) {
-        guard previousFlags != flags else { return }
-        previousFlags = flags
-
-        listener?(networkReachabilityStatusForFlags(flags))
-    }
+        let newStatus = NetworkReachabilityStatus(flags)
 
-    // MARK: - Internal - Network Reachability Status
+        guard previousStatus != newStatus else { return }
 
-    func networkReachabilityStatusForFlags(_ flags: SCNetworkReachabilityFlags) -> NetworkReachabilityStatus {
-        guard isNetworkReachable(with: flags) else { return .notReachable }
+        previousStatus = newStatus
```

Comment:
Do we also need to lock status read access inside the reachability queue to be fully thread-safe?

Reply:
I've made `previousStatus` protected in https://github.com/Alamofire/Alamofire/pull/2915/commits/005376d2c1714e377f9e6064205b91eefaee34a3, which I think is enough to protect our usage. I'm not sure about the thread-safety of `reachability` instance itself, but our usage should be fine. Let me know what you think.

Reply:
Seems good to me as long as `SCNetworkReachabilityGetFlags` is thread-safe since we're regenerating status each time we call it.

Reply:
We don't ever call this API directly, only users will, by calling one of the `is` properties, or getting the `status` or `flags` directly. Our usage of `flags` comes from the notification callback on the background queue, which only accesses the `previousStatus` value when calling the `listener`